scikeras/utils/transformers.py

```diff
@@ -0,0 +1,197 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Union
+
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, OrdinalEncoder
+from sklearn.utils.multiclass import type_of_target
+from tensorflow.python.keras.losses import is_categorical_crossentropy
+
+
+class Ensure2DTransformer(TransformerMixin, BaseEstimator):
+    """Transforms from 1D -> 2D and back.
+    """
+
+    def fit(self, X: np.ndarray) -> "Ensure2DTransformer":
+        self.should_transform_ = X.ndim == 1
```

Comment:
Style nit: can't this class be a lot simpler?

``` python
class Ensure2DTransformer:
    def fit(self, X):
        assert X.ndim <= 2
        self.ndim_ = X.ndim
        self.shape_ = (-1, 1) if X.ndim == 1 else (-1, X.shape[1])
        return self

    def transform(self, X):
        if X.ndim > 2:
            raise ValueError(...)
        return X.reshape(self.shape_)

    def inverse_transform(self, X):
        if self.ndim_ == 1:
            return np.squeeze(X, axis=1)
        return X
```

Reply:
That's basically what it currently is: https://github.com/adriangb/scikeras/blob/9f633f5c21a45b43986a8d779c3ef9e3005038bb/scikeras/_utils.py#L62
At some point in this PR we decided it needed the extra complexity, but perhaps that was because of some other part of the implementation at that point in time.

I checked and we can do that and still pass all tests (except the ones that were specifically checking for ValueErrors). Why not do:
``pyhon3
class Ensure2DTransformer(TransformerMixin, BaseEstimator):
    def fit(self, X: np.ndarray) -> "Ensure2DTransformer":
        return self

    def transform(self, X: np.ndarray) -> np.ndarray:
        if X.ndim == 1:
            X = X.reshape(-1, 1)
        return X

    def inverse_transform(self, X: np.ndarray) -> np.ndarray:
        if X.ndim == 2:
            X = np.squeeze(X, axis=1)
        return X
```

Seems even simpler.

Reply:
This has the advantage of handling almost-2D arrays. This code passes:

``` python
x = np.random.uniform(size=(100, 10, 1, 1, 1, 1))
assert x.reshape(100, 10).shape == (100, 10)
```

Reply:
Oh I see, you were preserving the `only inverse transform if the input was 1D` part.

Reply:
It currently fails for `(n_samples, 1)` inputs because it gets squeezed before recording that it was a 2D input. We want 2D in -> 2D out. Easy fix though.

Reply:
I also don't know that we need to generalize to higher dimensions... This transformer specifically exists to facilitate the use of feature encoders (`OneHotEncoder`, `OrdinalEncoder`) which require a 2D input with targets (which are often 1D). I think it's functionality should be as simple as "if the input is 1D, make it 2D and convert it back to 1D. If it's any other dimensionality, pass it through."

Reply:
That's a good idea I think. It's very simple documentation too.

But then the name `Ensure2DTransformer` is inaccurate. How 'bout `LabelReshaper`?

Reply:
Sure, I'll edit the name